src/transport/SecureSessionMgr.cpp

```diff
@@ -92,136 +92,154 @@ CHIP_ERROR SecureSessionMgr::SendMessage(NodeId peerNodeId, System::PacketBuffer
-            SuccessOrExit(err);
-
-            err = state->GetSecureSession().Encrypt(data, totalLen, data, packetHeader, mac);
-            SuccessOrExit(err);
-            err = mac.Encode(packetHeader, &data[totalLen], kMaxTagLen, &taglen);
-            SuccessOrExit(err);
+        err = EncryptMessage(state, payloadHeader, packetHeader, peerNodeId, msgBuf);
+        SuccessOrExit(err);
+    }
 
-            VerifyOrExit(CanCastTo<uint16_t>(totalLen + taglen), err = CHIP_ERROR_INTERNAL);
-            msgBuf->SetDataLength(static_cast<uint16_t>(totalLen + taglen));
+    // The start of buffer points to the beginning of the encrypted header, and the length of buffer
+    // contains both the encrypted header and encrypted data.
+    // Locally store the start and length of the retained buffer after accounting for the size of packet header.
+    headerSize = packetHeader.EncodeSizeBytes();
 
-            ChipLogDetail(Inet, "Secure transport transmitting msg %u after encryption", msgId);
-        }
+    msgStart = static_cast<uint8_t *>(msgBuf->Start() - headerSize);
+    msgLen   = static_cast<uint16_t>(msgBuf->DataLength() + headerSize);
 
-        if (bufferRetainSlot)
-        {
-            // The start of buffer points to the beginning of the encrypted header, and the length of buffer
-            // contains both the encrypted header and encrypted data.
-            // Locally store the start and length of the retained buffer after accounting for the size of packet header.
-            headerSize = packetHeader.EncodeSizeBytes();
+    // Retain the PacketBuffer in case it's needed for retransmissions.
+    encryptedMsg        = msgBuf.Retain();
+    encryptedMsg.mMsgId = packetHeader.GetMessageId();
 
-            p   = static_cast<uint8_t *>(msgBuf->Start() - headerSize);
-            len = static_cast<uint16_t>(msgBuf->DataLength() + headerSize);
+    ChipLogProgress(Inet, "Sending msg from %llu to %llu", mLocalNodeId, peerNodeId);
 
-            // Retain the PacketBuffer for following retransmit.
-            *bufferRetainSlot        = msgBuf.Retain();
-            bufferRetainSlot->mMsgId = msgId;
-        }
+    err = mTransportMgr->SendMessage(packetHeader, state->GetPeerAddress(), std::move(msgBuf));
+    SuccessOrExit(err);
 
-        err = mTransportMgr->SendMessage(packetHeader, state->GetPeerAddress(), std::move(msgBuf));
+    if (bufferRetainSlot != nullptr)
+    {
+        // Rewind the start and len of the buffer back to pre-send state for following possible retransmition.
+        encryptedMsg->SetStart(msgStart);
+        encryptedMsg->SetDataLength(msgLen);
 
-        if (bufferRetainSlot)
-        {
-            // Rewind the start and len of the buffer back to pre-send state for following possible retransmition.
-            (*bufferRetainSlot)->SetStart(p);
-            (*bufferRetainSlot)->SetDataLength(len);
-        }
+        (*bufferRetainSlot) = encryptedMsg.Retain();
```

Comment:
encryptedMsg.Retain() return PacketBufferHandle instead of EncryptedPacketBufferHandle, so it will call the following move constructor:

EncryptedPacketBufferHandle(PacketBufferHandle && aBuffer) : PacketBufferHandle(std::move(aBuffer)), mMsgId(0) {}

We still need to copy the MsgId manually in this case, otherwise, the outgoing packet will not be removed after it received the ack since its MgsId does not match. 

Suggested Change:
(*bufferRetainSlot) = encryptedMsg.Retain();
bufferRetainSlot->mMsgId = encryptedMsg.mMsgId;